File: src/lib/apiErrors.ts
import { toast } from "sonner";

export type ApiErrorType =
  | "API_KEY_EXHAUSTED"
  | "API_KEY_INVALID"
  | "API_RATE_LIMIT"
  | "API_KEY_MISSING"
  | "API_PERMISSION_DENIED"
  | "INVALID_MODEL"
  | "PROVIDER_UNREACHABLE"
  | "PROVIDER_ERROR";

export interface ApiErrorPayload {
  error: true;
  error_type: ApiErrorType;
  message: string;
}

export const API_KEY_USER_MESSAGE =
  "API token limit reached. Please change or update the API key.";

export class ApiClientError extends Error {
  readonly status: number;
  readonly apiError?: ApiErrorPayload;

  constructor(message: string, status: number, apiError?: ApiErrorPayload) {
    super(message);
    this.name = "ApiClientError";
    this.status = status;
    this.apiError = apiError;
  }
}

export function isApiKeyError(err: unknown): err is ApiClientError {
  return err instanceof ApiClientError && !!err.apiError;
}

export function isApiKeyErrorType(type: ApiErrorType | undefined): boolean {
  return (
    type === "API_KEY_EXHAUSTED" ||
    type === "API_KEY_INVALID" ||
    type === "API_RATE_LIMIT" ||
    type === "API_KEY_MISSING"
  );
}

function parseJsonBody(text: string): Record<string, unknown> | null {
  try {
    return JSON.parse(text) as Record<string, unknown>;
  } catch {
    return null;
  }
}

export function parseApiErrorFromBody(
  text: string,
  status: number
): ApiClientError | null {
  const json = parseJsonBody(text);
  if (!json) return null;

  if (json.error === true && typeof json.error_type === "string") {
    const payload: ApiErrorPayload = {
      error: true,
      error_type: json.error_type as ApiErrorType,
      message: String(json.message ?? API_KEY_USER_MESSAGE),
    };
    return new ApiClientError(payload.message, status, payload);
  }

  const detail = json.detail;
  if (typeof detail === "object" && detail !== null && !Array.isArray(detail)) {
    const d = detail as Record<string, unknown>;
    if (d.error === true && typeof d.error_type === "string") {
      const payload: ApiErrorPayload = {
        error: true,
        error_type: d.error_type as ApiErrorType,
        message: String(d.message ?? API_KEY_USER_MESSAGE),
      };
      return new ApiClientError(payload.message, status, payload);
    }
  }

  return null;
}

export async function throwIfApiError(res: Response): Promise<void> {
  if (res.ok) return;
  const text = await res.text();
  const parsed = parseApiErrorFromBody(text, res.status);
  if (parsed) throw parsed;
  throw new ApiClientError(text || `Request failed: ${res.status}`, res.status);
}

export function maskApiKey(key: string): string {
  const k = key.trim();
  if (!k) return "";
  if (k.length <= 10) return `${k.slice(0, 3)}****`;
  return `${k.slice(0, 6)}...${k.slice(-4)}`;
}

export function getUserFacingApiMessage(err: unknown): string {
  if (isApiKeyError(err)) {
    const type = err.apiError!.error_type;
    const msg = err.apiError!.message;
    if (type === "API_KEY_EXHAUSTED") {
      return "Your API key is valid, but quota or token limit is exhausted.";
    }
    if (type === "API_RATE_LIMIT") {
      return "Your API key is likely valid, but Groq rate limit was hit. Try again shortly.";
    }
    if (type === "API_KEY_INVALID") {
      return "Groq rejected this API key. Check GROQ_API_KEY in .env (no spaces/quotes) and restart the backend.";
    }
    if (type === "API_KEY_MISSING") {
      return "No API key configured. Add GROQ_API_KEY to your backend .env file and restart the server.";
    }
    if (type === "INVALID_MODEL") {
      return msg || "The configured Groq model name is invalid or unavailable.";
    }
    if (type === "API_PERMISSION_DENIED") {
      return msg || "API key is valid but lacks permission for this operation.";
    }
    if (type === "PROVIDER_UNREACHABLE") {
      return "Cannot reach Groq. Check network and that the backend is running.";
    }
    if (type === "PROVIDER_ERROR") {
      return msg || "Groq returned an error. See backend logs for details.";
    }
    return msg;
  }
  if (err instanceof Error && /failed to fetch|network/i.test(err.message)) {
    return "Backend is not reachable. Start the FastAPI server and check VITE_API_BASE.";
  }
  return err instanceof Error ? err.message : "Request failed";
}

export function notifyApiError(err: unknown): void {
  const message = getUserFacingApiMessage(err);
  const duration = isApiKeyError(err) ? 12_000 : 5_000;
  toast.error(message, { duration });
}

export type ApiKeyConnectionStatus =
  | "connected"
  | "missing"
  | "invalid"
  | "quota_exhausted"
  | "unknown";

export interface ApiKeyStatusResponse {
  status: string;
  error_type?: ApiErrorType;
  message?: string;
  key_count?: number;
  provider?: string;
  model?: string;
  masked_key?: string;
}

export interface ApiKeyValidateResponse {
  valid: boolean;
  error_type?: ApiErrorType;
  message: string;
  provider?: string;
  model?: string;
  masked_key?: string;
  preview?: string;
}

export function mapStatusToIndicator(
  data: ApiKeyStatusResponse | null,
  lastErrorType?: ApiErrorType
): ApiKeyConnectionStatus {
  if (lastErrorType === "API_KEY_EXHAUSTED" || lastErrorType === "API_RATE_LIMIT") {
    return "quota_exhausted";
  }
  if (lastErrorType === "API_KEY_INVALID" || lastErrorType === "API_PERMISSION_DENIED") {
    return "invalid";
  }
  if (lastErrorType === "INVALID_MODEL") return "unknown";
  if (!data) return "unknown";
  if (data.status === "missing") return "missing";
  if (data.status === "configured" && (data.key_count ?? 0) > 0) return "connected";
  return "unknown";
}

File: agents/api_errors.py
"""
Classify Groq / LLM failures into structured API-key errors for the FastAPI layer.
Uses provider SDK types first; avoids broad substring false positives.
"""

from __future__ import annotations

import re
from typing import Any, Optional

# Default Groq chat model for this project
GROQ_DEFAULT_MODEL = "llama-3.3-70b-versatile"


class ApiKeyError(Exception):
    """Raised when an LLM call fails due to API key, quota, model, or provider limits."""

    def __init__(self, error_type: str, message: str, status_code: int = 402):
        super().__init__(message)
        self.error_type = error_type
        self.message = message
        self.status_code = status_code


def mask_key(key: str | None) -> str:
    if not key:
        return ""
    k = key.strip()
    if len(k) <= 10:
        return "gsk_****"
    return f"{k[:6]}...{k[-4:]}"


def normalize_api_key(key: str | None) -> str:
    if not key:
        return ""
    k = key.strip()
    if (k.startswith('"') and k.endswith('"')) or (k.startswith("'") and k.endswith("'")):
        k = k[1:-1].strip()
    return k


def _exception_text(exc: Exception) -> str:
    parts: list[str] = [str(exc)]
    for attr in ("message", "body", "text"):
        val = getattr(exc, attr, None)
        if val is not None:
            parts.append(str(val))
    response = getattr(exc, "response", None)
    if response is not None:
        try:
            parts.append(str(getattr(response, "text", "") or ""))
        except Exception:
            pass
    status = getattr(exc, "status_code", None)
    if status is not None:
        parts.append(str(status))
    return " ".join(parts).lower()


def _status_code(exc: Exception) -> int | None:
    code = getattr(exc, "status_code", None)
    if isinstance(code, int):
        return code
    response = getattr(exc, "response", None)
    if response is not None:
        sc = getattr(response, "status_code", None)
        if isinstance(sc, int):
            return sc
    return None


def _is_rate_limit_text(text: str) -> bool:
    return (
        "429" in text
        or "rate_limit" in text
        or "rate limit" in text
        or "too many requests" in text
        or "rate_limit_exceeded" in text
    )


def _is_invalid_key_text(text: str) -> bool:
    """Strict patterns — do not match generic 'authentication' in prompts."""
    patterns = (
        r"invalid\s+api\s+key",
        r"invalid_api_key",
        r"incorrect\s+api\s+key",
        r"invalid\s+x-api-key",
        r"api\s+key\s+is\s+invalid",
        r"authentication\s+failed",
        r"auth(?:entication)?\s+error.*api\s+key",
    )
    return any(re.search(p, text) for p in patterns)


def _is_model_error_text(text: str) -> bool:
    if "model" not in text:
        return False
    return any(
        m in text
        for m in (
            "not found",
            "does not exist",
            "decommissioned",
            "deprecated",
            "unknown model",
            "model_not_found",
            "invalid model",
        )
    )


def _classify_groq_sdk(exc: Exception) -> Optional[ApiKeyError]:
    try:
        from groq import APIStatusError, AuthenticationError, PermissionDeniedError, RateLimitError
    except ImportError:
        return None

    if isinstance(exc, AuthenticationError):
        return ApiKeyError(
            "API_KEY_INVALID",
            "The API key was rejected by Groq. Check GROQ_API_KEY in .env (no extra spaces or quotes).",
            401,
        )
    if isinstance(exc, PermissionDeniedError):
        return ApiKeyError(
            "API_PERMISSION_DENIED",
            "The API key is valid but lacks permission for this operation.",
            403,
        )
    if isinstance(exc, RateLimitError):
        return ApiKeyError(
            "API_RATE_LIMIT",
            "API rate limit exceeded. Your key is likely valid — wait and retry or use another key.",
            429,
        )
    if isinstance(exc, APIStatusError):
        status = _status_code(exc) or 500
        text = _exception_text(exc)
        if status == 401 and _is_invalid_key_text(text):
            return ApiKeyError(
                "API_KEY_INVALID",
                "The API key was rejected by Groq.",
                401,
            )
        if status == 429 or _is_rate_limit_text(text):
            return ApiKeyError(
                "API_RATE_LIMIT",
                "API rate limit exceeded. Your key may still be valid.",
                429,
            )
        if _is_model_error_text(text) or (status == 404 and "model" in text):
            return ApiKeyError(
                "INVALID_MODEL",
                f"The model '{GROQ_DEFAULT_MODEL}' is unavailable. Check Groq model name in backend config.",
                400,
            )
        if status in (402, 403) and any(
            m in text for m in ("quota", "billing", "credit", "insufficient")
        ):
            return ApiKeyError(
                "API_KEY_EXHAUSTED",
                "API key is valid, but quota or billing limit is exhausted.",
                402,
            )
    return None


def classify_llm_error(exc: Exception) -> Optional[ApiKeyError]:
    """Map provider exceptions to ApiKeyError, or return None if unrelated."""
    if isinstance(exc, ApiKeyError):
        return exc

    sdk = _classify_groq_sdk(exc)
    if sdk:
        return sdk

    text = _exception_text(exc)
    status = _status_code(exc)

    if "no groq_api_key" in text or ("not configured" in text and "groq" in text):
        return ApiKeyError(
            "API_KEY_MISSING",
            "No GROQ_API_KEY configured. Set GROQ_API_KEY in the project root .env and restart the backend.",
            503,
        )

    if _is_model_error_text(text):
        return ApiKeyError(
            "INVALID_MODEL",
            f"Model error from Groq (configured: {GROQ_DEFAULT_MODEL}).",
            400,
        )

    if status == 429 or _is_rate_limit_text(text):
        return ApiKeyError(
            "API_RATE_LIMIT",
            "API rate limit exceeded. Your key may still be valid — try again later.",
            429,
        )

    if any(
        m in text
        for m in (
            "quota",
            "insufficient_quota",
            "billing",
            "credit",
            "exhausted",
            "usage limit",
            "tokens per day",
            "tokens per minute",
        )
    ):
        return ApiKeyError(
            "API_KEY_EXHAUSTED",
            "API key is valid, but quota or token limit is exhausted.",
            402,
        )

    if _is_invalid_key_text(text):
        return ApiKeyError(
            "API_KEY_INVALID",
            "The API key was rejected by Groq.",
            401,
        )

    if status == 401 and any(m in text for m in ("invalid", "unauthorized", "api key")):
        return ApiKeyError(
            "API_KEY_INVALID",
            "The API key was rejected by Groq.",
            401,
        )

    if any(m in text for m in ("connection", "timeout", "network", "refused", "unreachable")):
        return ApiKeyError(
            "PROVIDER_UNREACHABLE",
            "Could not reach Groq. Check network and backend connectivity.",
            503,
        )

    return None


def reraise_if_api_key_error(exc: Exception) -> None:
    """Re-raise as ApiKeyError when the underlying failure is key/quota related."""
    if isinstance(exc, ApiKeyError):
        raise exc
    classified = classify_llm_error(exc)
    if classified:
        raise classified from exc
